File: python-algorithm/leetcode/problem_45.py
"""45. Jump Game II
https://leetcode.com/problems/jump-game-ii/
"""
from typing import List


class Solution:
    def jump2(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [0] * n
        for i in range(n - 2, -1, -1):
            if i + nums[i] >= n - 1:
                steps = n + 1
                for j in range(1, nums[i] + 1):
                    pass
        return dp[0]

    def jump(self, nums: List[int]) -> int:
        def test(i: int, pre_steps: int):
            if i == size - 1:
                return
            if memo[-1] < size:  # it means already has jumped to last.
                if pre_steps + 1 >= memo[-1]:
                    return
            for j in range(nums[i], 0, -1):
                k = min(i + j, size - 1)
                if memo[k] < size:  # it means jump to nums[k] before.
                    if pre_steps + 1 < memo[k]:  # then we can try to jump.
                        memo[k] = min(pre_steps + 1, memo[k])
                        test(k, pre_steps + 1)
                else:
                    memo[k] = pre_steps + 1
                    test(k, pre_steps + 1)

        size = len(nums)
        if size <= 1:
            return 0
        memo = [size] * size  # memo[i] = the least jumps to the nums[i].
        test(0, 0)
        return memo[-1]

File: java-algo/src/main/java/com/leetcode/Problem45.java
package com.leetcode;

/**
 * 45. Jump Game II
 * https://leetcode.com/problems/jump-game-ii/
 *
 * Given an array of non-negative integers, you are initially positioned
 * at the first index of the array.
 *
 * Each element in the array represents your maximum jump length at that position.
 *
 * Your goal is to reach the last index in the minimum number of jumps.
 *
 * Example:
 *
 * Input: [2,3,1,1,4]
 * Output: 2
 * Explanation: The minimum number of jumps to reach the last index is 2.
 *     Jump 1 step from index 0 to 1, then 3 steps to the last index.
 * Note:
 *
 * You can assume that you can always reach the last index.
 */
public class Problem45 {

    public int jump(int[] nums) {
        int size = nums.length;
        if (size < 2)
            return 0;
        int[] memo = new int[size];
        for (int i = 0; i < size; i++)
            memo[i] = size;
        try_jump(0, 0, nums, memo);
        return memo[size - 1];
    }

    private void try_jump(int i, int pre_steps, int[] nums, int[] memo) {
        if (i == nums.length - 1)
            return;
        for (int j = nums[i]; j >= 0; j--) {
            int k = Math.min(i + j, nums.length - 1);
            if (memo[k] < nums.length) {
                if (pre_steps + 1 < memo[k]) {
                    memo[k] = Math.min(pre_steps + 1, memo[k]);
                    try_jump(k, pre_steps + 1, nums, memo);
                }
            } else {
                memo[k] = pre_steps + 1;
                try_jump(k, pre_steps + 1, nums, memo);
            }
        }
    }
}
